pkg/controllers/node/ipam.go

```diff
@@ -217,24 +239,56 @@ func (c *NodeController) nodeExists(node string) bool {
 	return true
 }
 
+// podExists returns whether the given pod exists in the Kubernetes API and is on the provided Kubernetes node.
+// Note that the "node" parameter is the name of the Kubernetes node in the Kubernetes API.
 func (c *NodeController) podExists(name, ns, node string) bool {
-	n, err := c.k8sClientset.CoreV1().Pods(ns).Get(name, v1.GetOptions{})
+	p, err := c.k8sClientset.CoreV1().Pods(ns).Get(name, v1.GetOptions{})
 	if err != nil {
 		if errors.IsNotFound(err) {
 			return false
 		}
 		log.WithError(err).Warn("Failed to query pod, assume it exists")
 	}
-	if n.Spec.NodeName != node {
+	if p.Spec.NodeName != node {
```

Comment:
Prolly just not grokking correctly, but for borrowed IPs won't this be the case and in that case we'll clean up the allocation.